src/cargo/core/compiler/compilation.rs

```diff
@@ -527,3 +527,98 @@ fn target_linker(bcx: &BuildContext<'_, '_>, kind: CompileKind) -> CargoResult<O
     }
     Ok(matching_linker.map(|(_k, linker)| linker.val.clone().resolve_program(bcx.gctx)))
 }
+
+/// This tracks environment variables Cargo sets to rustc when building a crate.
+///
+/// This only inclues variables with statically known keys.
+/// For environment variable keys that vary like `CARG_BIN_EXE_<name>` or
+/// `-Zbindeps` related env vars, we compare only their prefixes to determine
+/// if they are internal.
+/// See [`is_env_set_by_cargo`] and
+/// <https://doc.rust-lang.org/nightly/cargo/reference/environment-variables.html#environment-variables-cargo-sets-for-crates>.
+const ENV_VARS_SET_FOR_CRATES: [&str; 23] = [
+    crate::CARGO_ENV,
+    "CARGO_MANIFEST_DIR",
+    "CARGO_MANIFEST_PATH",
+    "CARGO_PKG_VERSION",
+    "CARGO_PKG_VERSION_MAJOR",
+    "CARGO_PKG_VERSION_MINOR",
+    "CARGO_PKG_VERSION_PATCH",
+    "CARGO_PKG_VERSION_PRE",
+    "CARGO_PKG_AUTHORS",
+    "CARGO_PKG_NAME",
+    "CARGO_PKG_DESCRIPTION",
+    "CARGO_PKG_HOMEPAGE",
+    "CARGO_PKG_REPOSITORY",
+    "CARGO_PKG_LICENSE",
```

Comment:
To remove these requires that we are tracking these somewhere else. For things like the package name, thats more obvious. For ones like this, where is that happening? Should we have a test that changing the license causes a rebuild?

Reply:
That's a good catch. We actually only rebuild for `authors`, `homepage`, `repository`, and `repository`.

https://github.com/rust-lang/cargo/blob/298e403317d231892e11bf93d3073f011841be48/src/cargo/core/compiler/fingerprint/mod.rs?plain=1#L1534

If we're going to hand-pick some of them, I would rather drop this out.

Since we were trying to fix #14811, the alternative is a “wontfix” and documenting it. 

> If you're setting `CARGO_SOMETHING` in `[env]`, you are screwed because it won't work and may hurt your rebuild detection.

Reply:
Apparently, `package.description` was tracked by fingerprint because of the usage of `CARGO_PKG_DESCRIPTION` <https://github.com/rust-lang/cargo/pull/3857>.

Okay, maybe the other way round—We do hand-pick whatever we want to "remove" from cargo's dep-info file. The other stay in dep-info so they are happy to rebuild when requires by `env!`. And we don't even need to track them in fingerprint.

Reply:
I think it makes sense for us to track metadata in dep info rather than fingerprint so we only rebuild if the user cares.

As for the other part, I'm unusre

Reply:
https://github.com/rust-lang/cargo/blob/3cd245ce48ee2646509f589d06b341334db8d9cd/src/cargo/core/manifest.rs?plain=1#L145-L146

It is so fun that `package.links` and `package.rust-version` are considered metadata and and won't trigger any rebuild when changed.

Reply:
imo there isn't a reason to rebiild on `rust-version` change unless the user is using it.

Reply:
That is fair. It is a kind of metadata for the actual compilation.